mailet/standard/src/test/java/org/apache/james/transport/matchers/AtMostTest.java

```diff
@@ -59,100 +61,199 @@ void setup() throws MessagingException {
+        void shouldThrowWithInvalidCondition() {
+            FakeMatcherConfig matcherConfig = FakeMatcherConfig.builder()
+                .matcherName("AtMost")
+                .condition("invalid")
+                .build();
+
+            assertThatThrownBy(() -> matcher.init(matcherConfig))
+                .isInstanceOf(MessagingException.class);
+        }
+
+        @Test
+        void shouldThrowWithNegativeCondition() {
+            FakeMatcherConfig matcherConfig = FakeMatcherConfig.builder()
+                .matcherName("AtMost")
+                .condition("-1")
+                .build();
+
+            assertThatThrownBy(() -> matcher.init(matcherConfig))
+                .isInstanceOf(MessagingException.class);
+        }
+
+        @Test
+        void shouldThrowWithConditionToZero() {
+            FakeMatcherConfig matcherConfig = FakeMatcherConfig.builder()
+                .matcherName("AtMost")
+                .condition("0")
+                .build();
+
+            assertThatThrownBy(() -> matcher.init(matcherConfig))
+                .isInstanceOf(MessagingException.class);
+        }
     }
 
-    @Test
-    void shouldMatchWhenNoRetries() throws MessagingException {
-        Mail mail = createMail();
-        mail.setAttribute(new Attribute(AT_MOST_EXECUTIONS, AttributeValue.of(0)));
-
-        Collection<MailAddress> actual = matcher.match(mail);
-
-        assertThat(actual).containsOnly(RECIPIENT1);
+    @Nested
+    class MultipleMatchersConfigurationTest {
+        private AtMost atMost2;
+        private AtMost atMost3;
+
+        @BeforeEach
+        void setup() throws MessagingException {
+            this.atMost2 = new AtMost();
+            atMost2.init(
+                FakeMatcherConfig.builder()
+                    .matcherName("AtMost")
+                    .condition("AtMost2:2")
+                    .build());
+
+            this.atMost3 = new AtMost();
+            atMost3.init(
+                FakeMatcherConfig.builder()
+                    .matcherName("AtMost")
+                    .condition("AtMost3:2")
+                    .build());
+        }
+
+        @Test
+        void matchersShouldStopWhenAMatcherReachedLimit() throws MessagingException {
+            Mail mail1 = createMail();
+
+            SoftAssertions.assertSoftly(Throwing.consumer(
+                softly -> {
+                    softly.assertThat(atMost2.match(mail1)).containsOnly(RECIPIENT1);
+                    softly.assertThat(atMost2.match(mail1)).containsOnly(RECIPIENT1);
+                    softly.assertThat(atMost2.match(mail1)).isEmpty();
+                    softly.assertThat(atMost3.match(mail1)).containsOnly(RECIPIENT1);
+                    softly.assertThat(atMost3.match(mail1)).containsOnly(RECIPIENT1);
+                    softly.assertThat(atMost3.match(mail1)).isEmpty();
+                }));
+        }
+
+        @Test
+        void matchersShouldMatchWhenLimitNotReached() throws MessagingException {
```

Comment:
Does this test makes sense in the mutliple matchers test nested class, as we only test a matcher here?

Reply:
Yes, think it tested by single matcher, i will delete it.